### PR TITLE
fix(22903): prevent empty dashboard pages when posts are archived

### LINKED ISSUE #22903: Archived posts result in empty pages on the posts dashboard
**Describe the bug**

When you archive a post on the dashboard, it hides it so that it is not displayed, but it still takes up a spot on the list. This means if you archive a lot of posts, you will have "empty" pages where no posts are shown and you need to go to another page to see the unarchived posts.

**To Reproduce**

1. Go to your posts dashboard: https://dev.to/dashboard
2. Archive many of your posts

You will notice the number of posts shown on the page will decrease as you archive posts. Since a page shows 25 posts by default, if you archive your first 25 posts, you will have an empty page where no posts are shown, and you'll need to move to page 2 to see your unarchived posts.

**Expected behavior**

Regardless of how many posts are archived, you should always see 25 unarchived posts on each page.

**Screenshots**

Image

**Desktop (please complete the following information):**

- OS, version: Windows 11
- Browser, version: Firefox v148.0

**Additional context**

The reason I am archiving so many posts is I am using the RSS extension (https://dev.to/settings/extensions) to import all of my posts from my blog. For old posts that I do not want to publish on dev.to, if I delete them, the RSS extension repopulates them in the dashboard later. So I am archiving the posts that I do not want to publish as a workaround.

### PR DESCRIPTION
**Pull request by anchildress1**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Bug Fix
- [ ] Feature
- [ ] Optimization
- [ ] Documentation Update

## Description

Archiving posts on the dashboard was hiding them with CSS but the server still counted them toward pagination. Archive 25 posts and page 1 is completely empty — your unarchived posts are on page 2. This hit hard for anyone using the RSS importer who archives old posts in bulk to prevent re-imports.

The fix moves filtering server-side so archived posts are excluded from the query entirely before pagination runs. The old `archivedPostFilters.js` client toggle is gone. Sorting and view-switching now preserve `show_archived` in the URL so the state doesn't vanish mid-navigation. After archiving/unarchiving, the page does a quick InstantClick refresh instead of trying to sync DOM state by hand — that DOM-sync code was brittle and not worth maintaining.

There's a couple of minor remaining edge cases visible in the recording below — they're not regressions and can be addressed separately.

1. Switching the selected filter while archived posts are visible can behave unexpectedly in some sequences.
2. ~~Navigating from page 2 of a view to a view that has only one page can leave the page looking broken.~~ Copilot found root cause and fixed already.

## Related Tickets & Documents

- Closes #22903

## QA Instructions, Screenshots, Recordings

1. Log in as a user with a mix of published and archived posts.
2. Visit `/dashboard` — archived posts should **not** appear, and the "Show archived" button should be hidden if none are archived, or visible (unpressed) if some exist.
3. Click "Show archived" — archived posts appear, URL gains `show_archived=true`, button reads "Hide archived" (`aria-pressed="true"`).
4. With archived posts visible, change the sort dropdown — confirm `show_archived=true` is preserved in the new URL and archived posts remain visible.
5. Click "Show quickie posts" or "Show full posts" — confirm `show_archived` is preserved.
6. Archive a post — page refreshes via InstantClick, post is no longer in the listing (or is, if `show_archived=true`).
7. Unarchive a post while `show_archived=true` — page refreshes, post reappears in the default listing.
8. Paginate to page 2+ and repeat steps 3–7.

https://github.com/user-attachments/assets/657a5b1c-5a9c-40ed-a9e7-288d30468637

### UI accessibility checklist
- [x] Semantic HTML implemented?
- [x] Keyboard operability supported?
- [ ] Checked with [axe DevTools](https://www.deque.com/axe/) and addressed `Critical` and `Serious` issues?
- [ ] Color contrast tested?

## Added/updated tests?

- [x] Yes
  - **RSpec request specs** (`spec/requests/dashboard_spec.rb`): +232 lines total across all three commits — archived/non-archived filtering, `show_archived` case sensitivity, pagination edge cases, status vs. full-post filtering, and archive/unarchive regression coverage.
  - **Jest unit tests** (`dashboardDropdowns.test.js`, `initializeDashboardSort.test.js`): updated for InstantClick refresh behavior; full coverage of `buildNavigationUrl` including param inheritance, explicit overrides, and edge cases (empty params, hash fragments).

## [optional] Are there any post deployment tasks we need to perform?

None.

### CODE REVIEW

File: app/javascript/packs/initializers/initializeDashboardSort.js [hunk 1]
```diff
@@ -1,3 +1,16 @@
...
+  const currentSearchParams = new URLSearchParams(window.location.search);
+
+  currentSearchParams.forEach((value, key) => {
+    if (!destination.searchParams.has(key)) {
+      destination.searchParams.set(key, value);
+    }
+  });
+
+  return `${destination.pathname}${destination.search}${destination.hash}`;
+}
```
[the remark this answers is not in the document]
Comment (anchildress1) [reply app/javascript/packs/initializers/initializeDashboardSort.js:12]: Valid. Inheriting `page` when changing sort or switching dashboard sections is one of the edge cases noted in the PR description — this resolves it. Fixed `buildNavigationUrl` to explicitly drop `page` from inherited params so navigation context changes always reset to page 1. Added a regression test covering this case; all 13 tests pass.

File: app/controllers/dashboards_controller.rb [hunk 2]
```diff
@@ -167,6 +169,12 @@ def feed_imports
 
   private
 
+  def filter_archived_articles
+    @show_archived = params[:show_archived] == "true"
+    @has_archived = @articles.exists?(archived: true)
+    @articles = @articles.where(archived: false) unless @show_archived
+  end
+
```
[the remark this answers is not in the document]
Comment (anchildress1) [reply app/controllers/dashboards_controller.rb:177]: Good catch. The suggestion to reuse `params[:filter]` would be architecturally messy — `filter` isn't used in the dashboard controller at all and conflating archived-state with post-type semantics would create confusion. The correct fix is to add `show_archived` to the Fastly safe params allowlist, which I've done in `config/fastly/snippets/safe_params_list.vcl`. Without that, the param would be stripped at the edge and the feature wouldn't work in production.

**Comment by anchildress1** [comment]
Closing as superseded by #23194, which shipped server-side archived filtering with a different design (`filter=archived` param, `@has_archived_articles`, inline filter in `show`). Both PRs solved issue #22903 — main's implementation is now live, so this branch's approach is no longer needed.
